### PR TITLE
Upgrade to Mypy 1.0

### PR DESCRIPTION
**Pull request by christianbundy**
# I have made things!

Upgrade to Mypy 1.0

## Related issues

Refs https://github.com/python/mypy/issues/13685

Closes #340

### CODE REVIEW

**Comment by intgr** [comment]
Approved CI run.

**Comment by intgr** [comment]
Some if not all of the typecheck errors are due to upstream changes in DRF test suite, not from upgrading mypy.

**Comment by christianbundy** [comment]
Do we just add them as excluded errors in `./scripts/typecheck_tests.py`? Trying to make sure I understand the procedure.

**Comment by intgr** [comment]
> Do we just add them as excluded errors in ./scripts/typecheck_tests.py?

I've looked at each error, and checked if it's something we're missing in stubs or a false positive. Usually it's a false positive, or some new feature that has been merged but will be in the next version (e.g. #413).

But it's annoying and confusing, especially for new contributors, when CI checks start to fail due to unrelated changes. So I've been thinking that we may want to freeze the git commit hash that we check (like is already done in django-stubs).

But the downside is that then we would have to periodically update this hash, something that's been a bit neglected in django-stubs.

**Comment by intgr** [comment]
I rebased your branch. The remaining CI errors are from mypy 1.0 update.

**Comment by christianbundy** [comment]
Thanks! For convenience, here are the relevant files:

- https://github.com/encode/django-rest-framework/blob/22d206c1e0dbc03840c4d190f7eda537c0f2010a/tests/test_permissions.py
- https://github.com/encode/django-rest-framework/blob/22d206c1e0dbc03840c4d190f7eda537c0f2010a/tests/test_versioning.py

## Problem 1

```
drf_source/tests/test_permissions.py:593: error: "<typing special form>" not callable
drf_source/tests/test_permissions.py:599: error: "<typing special form>" not callable
drf_source/tests/test_permissions.py:653: error: "<typing special form>" not callable
drf_source/tests/test_permissions.py:661: error: <nothing> not callable
drf_source/tests/test_permissions.py:673: error: "<typing special form>" not callable
drf_source/tests/test_permissions.py:681: error: <nothing> not callable
drf_source/tests/test_permissions.py:736: error: "<typing special form>" not callable
```

These seem to be exclusively complaining about this pattern:

```python3
composed_perm = permissions.IsAuthenticated | permissions.AllowAny
assert composed_perm().has_permission(request, None) is True
```

This only happens when you `__or__` permissions, and it looks like `__and__` is unaffected.

I don't really understand why this would happen, since [the stub definition](https://github.com/typeddjango/djangorestframework-stubs/blob/bf9d3837eaf0b1e2928f28bad48c1f617813376b/rest_framework-stubs/permissions.pyi#L18-L23) seems to treat `|` and `&` similarly, but I'm inclined to think that maybe this is a bug in Mypy where `|` is being treated as a type union. What do you think?

## Problem 2

```
drf_source/tests/test_versioning.py:28: error: "None" not callable
drf_source/tests/test_versioning.py:35: error: "None" not callable
drf_source/tests/test_versioning.py:43: error: "None" not callable
drf_source/tests/test_versioning.py:50: error: "None" not callable
```

These are all coming from:

```python3
self.versioning_class()
```

Where `self` is a subclass of `APIView` with a `get()` method defined. 

[The stub defines this as nullable](https://github.com/typeddjango/djangorestframework-stubs/blob/bf9d3837eaf0b1e2928f28bad48c1f617813376b/rest_framework-stubs/views.pyi#L54), and [DRF source](https://github.com/encode/django-rest-framework/blob/4abfa28e0879e2df45937ac8c7a9ffa161561955/rest_framework/views.py#L378) seems to agree that this is possible, so I can add an ignore for this one.

**Comment by intgr** [comment]
> I don't really understand why this would happen, since [the stub definition](https://github.com/typeddjango/djangorestframework-stubs/blob/bf9d3837eaf0b1e2928f28bad48c1f617813376b/rest_framework-stubs/permissions.pyi#L18-L23) seems to treat | and & similarly, but I'm inclined to think that maybe this is a bug in Mypy where | is being treated as a type union. What do you think?

Yes indeed, mypy thinks `composed_perm` is a type alias for Union. For now, an ignore is fine.

I always hated how DRF requires permissions checks to be classes, not instances. 

This can cause problems in actual user code. Maybe report a mypy bug? Mypy could look at whether a class overrides the `__or__` method, or look at the method's return value, to detect this.

File: scripts/typecheck_tests.py [hunk 1]
```diff
@@ -188,8 +188,6 @@
         "base class",
         '"CharField" has incompatible type "Collection[Any]"',
         'Name "foo" is not defined',
-        'Unsupported left operand type for | ("ReturnDict")',
-        'Unsupported left operand type for | ("Dict[str, str]")',
```
Comment (intgr) [inline scripts/typecheck_tests.py:192]: You can't remove these, they are still necessary for older Python versions.

**Review by intgr — CHANGES_REQUESTED** [review]
There are still some test failures. Sorry, I could have approved the CI run earlier.

File: scripts/typecheck_tests.py (hunk 1 above)
Comment (christianbundy) [reply scripts/typecheck_tests.py]: ah, I didn't realize that -- thanks!

**Comment by christianbundy** [comment]
Done! And I ran `act` locally and it looks like it passed all tests. (Though I'd _love_ to be able to run CI without approvals if that's easy to fix.)

**Comment by flaeppe** [comment]
> Done! And I ran `act` locally and it looks like it passed all tests. (Though I'd _love_ to be able to run CI without approvals if that's easy to fix.)

I've approved the CI run. You shouldn't need approvals once you've become a contributor.

**Review by intgr — APPROVED** [review]
Thanks!

**Pull request merged**